eventauth/eventauth.go

```diff
@@ -251,6 +229,73 @@ func redactEventAllowed(event Event, authEvents AuthEvents) error {
 	panic("Not implemented")
 }
 
+// defaultEventAllowed checks whether the event is allowed by the default
+// checks for events.
+// It returns an error if the event is not allowed or if there was a
+// problem loading the auth events needed.
 func defaultEventAllowed(event Event, authEvents AuthEvents) error {
-	panic("Not implemented")
+	var allower eventAllower
+	if err := allower.setup(authEvents, event.Sender); err != nil {
+		return err
+	}
+
+	return allower.commonChecks(event)
+}
+
+// An eventAllower has the information needed to authorise all events types
+// other than m.room.create, m.room.member and m.room.alias which are special.
+type eventAllower struct {
+	create      createContent
+	member      memberContent
+	powerLevels powerLevelContent
+}
+
+// setup loads the necesary information from the auth events.
+func (e *eventAllower) setup(authEvents AuthEvents, senderID string) error {
+	if err := e.create.load(authEvents); err != nil {
+		return err
+	}
+	if err := e.member.load(authEvents, senderID); err != nil {
+		return err
+	}
+	if err := e.powerLevels.load(authEvents, e.create.Creator); err != nil {
+		return err
+	}
+	return nil
+}
+
+// commonChecks does the checks that are applied to all events types other than
+// m.room.create, m.room.member, or m.room.alias.
+func (e *eventAllower) commonChecks(event Event) error {
+	if event.RoomID != e.create.roomID {
+		return errorf("create event has different roomID: %q != %q", event.RoomID, e.create.roomID)
+	}
+
+	if err := e.create.userIDAllowed(event.Sender); err != nil {
+		return err
+	}
+
+	if e.member.Membership != "join" {
+		return errorf("sender %q not in room", event.Sender)
```

Comment:
How does this work on transition states? E.g. invite -> join and join -> leave? Comment please.